Implement #2765: Document Azure Key Vault KMS usage for record encryption (#2894)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/record-encryption/azure-key-vault/proc-azure-key-vault-setup.adoc

:_mod-docs-content-type: PROCEDURE

// file included in the following:
//
// assembly-azure-key-vault.adoc

[id='proc-azure-key-vault-setup-{context}']
= Setting up Azure resources

> [role="_abstract"]

Reviewer: We still need an abstract intro for the procedure.

Maybe:

[role="_abstract"]
Use this procedure to prepare the Azure resources required for Kroxylicious envelope encryption with Azure Key Vault. You create a resource group, provision a key vault, configure a key management user, and establish an authentication method that the filter uses to access the key vault.

Author: ```suggestion
[role="_abstract"]
This procedure describes how to prepare the Azure resources required to use the Kroxylicious Record Encryption filter with Azure Key Vault. This process uses the Azure CLI to create a resource group, provision a key vault, configure a key management user, and establish an authentication method that the filter will use to access the key vault.
```

.Prerequisites

* An Azure subscription that includes Azure Key Vault
* An Azure user with sufficient permissions to create and manage users, resource groups, roles, and resources

.Procedure

. Create a resource group
+
Before a key vault can be created, you must first create a resource group into which your key vault will be deployed.
+
If you are deploying Kroxylicious into multiple different environments (for example, into your "development", "test", and "production" environments), you should have separate resource groups (and separate key vaults) for each environment.
+
NOTE: Resource groups are tied to the region they were created with. While resources (such as key vaults) within a resource group can be deployed to any region, Microsoft recommends that they should be located in the same region as the resource group. Keep this in mind when choosing a region for your resource groups.
+
If using the Azure CLI, you can create a resource group with a command like this:
+
[source,shell]
----
az group create --name "my-resource-group" --location "eastus"
----

> . Create a key vault

Author: We still need this to be a numbered list item, or else the numbering will reset on the subsequent line.
```suggestion
. Create a key vault with the Azure CLI:
+
[source,shell]
----
KEY_VAULT_NAME="kroxylicious-key-vault"
RESOURCE_GROUP_NAME="my-resource-group"
az keyvault create --name ${KEY_VAULT_NAME} --resource-group ${RESOURCE_GROUP_NAME}
----
* Key vault names must be unique.
* `RESOURCE_GROUP_NAME` should match the name of the resource group created in the previous section.
```

+
You can create a key vault with the Azure CLI using a command like this:
+
[source,shell]
----
KEY_VAULT_NAME="kroxylicious-key-vault"
RESOURCE_GROUP_NAME="my-resource-group"
az keyvault create --name ${KEY_VAULT_NAME} --resource-group ${RESOURCE_GROUP_NAME}
----
* Key vault names must be unique.
* `RESOURCE_GROUP_NAME` should match the name of the resource group created in the previous section.

. Create a key management user
+
To use the filter, a user such as the Key Vault Owner —or any user with the *Key Vault Crypto Officer* RBAC role or equivalent permissions on the key vault— must create the encryption keys within the key vault, which
are used by the xref:con-record-encryption-overview-{context}[envelope encryption] process.
+
Once a key vault is created, the only user able to manage the vault is the user that created it (the Key Vault Owner). Access to key vaults is managed using RBAC roles (with some exceptions, see the Microsoft documentation for details), which can be conferred by a user with a minimum of either Role Based Access Control Administrator role or User Access Administrator role.
+

> The following Azure CLI commands demonstrate creating a user in Microsoft Entra ID and granting key management RBAC privileges. This example illustrates using the domain `example.com` but you should use your own domain (e.g. a `*.onmicrosoft.com` domain) for this. You can also substitute the value in the `USER_NAME` field (`kroxylicious-user`) for a different username if preferred.

Author: ```suggestion
The following Azure CLI commands demonstrate creating a user in Microsoft Entra ID and granting key management RBAC privileges. This example illustrates using the domain `example.com` but you must use your own domain, such as a `*.onmicrosoft.com` domain.
Substitute the value in the `USER_NAME` field (`kroxylicious-user`) for a different username if preferred.
```

USER_PRINCIPAL_NAME="${USER_NAME}@${DEFAULT_DOMAIN}"
INITIAL_PASSWORD=$(tr -dc 'A-Za-z0-9!?%=' < /dev/urandom | head -c 16)

# Create user
az ad user create \
  --display-name "${USER_NAME}" \
  --user-principal-name "${USER_PRINCIPAL_NAME}" \
  --password "${INITIAL_PASSWORD}" \
  --force-change-password-next-sign-in true

# Create RBAC role assignment
az role assignment create \
  --role "Key Vault Crypto Officer" \
  --assignee ${USER_PRINCIPAL_NAME} \
  --scope "/subscriptions/${SUBSCRIPTION_ID}/resourceGroups/${RESOURCE_GROUP_NAME}/providers/Microsoft.KeyVault/vaults/${KEY_VAULT_NAME}"

echo "Now log in at ${PORTAL_URL}  with user name ${USER_PRINCIPAL_NAME} and password \"${INITIAL_PASSWORD}\" and change the password."
----
* You can view a list of all the subscriptions for your current logged-in user with `az account list` and change the active subscription with `az account set`.
* Your Azure portal URL will vary depending on your current Azure cloud.
* The Key Vault Crypto Officer role allows the user to perform any action on the keys within the key vault identified in the `--scope`, except manage permissions.

> * The scope in the role assignment command is the full scope hierarchy of your key vault. For more information on this format see the Microsoft docs.

Reviewer: ```suggestion
* View all subscriptions for the logged-in user with `az account list`, and change the active subscription with `az account set`.
* The Azure portal URL varies depending on your current Azure cloud.
* The Key Vault Crypto Officer role lets a user perform any action on the keys within the key vault identified in the `--scope`, except managing permissions.
* The scope in the role assignment command specifies the full scope hierarchy of the key vault. For more information on this format, see the Microsoft documentation.
```
